### PR TITLE
Implement a python wrapper around the patch_subset c++ code.

### PR DESCRIPTION
**Pull request by garretrieger**
Adds a python PFE method which uses the c++ patch subset code to perform PFE.

### CODE REVIEW

File: analysis/analyzer.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+]
+
+
+def cost(time_ms):
+  """Assigns a cost to a measure of request latency."""
+  # TODO(garretrieger): implement me.
+  return time_ms
+
+
+def analyze_data_set(data_set, pfe_methods, network_models, font_directory):
```
Comment (rsheeter) [inline analysis/analyzer.py:55]: enough parts a single type (maybe just a namedtuple) capturing all the input bits might make sense?

File: analysis/simulation.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+def simulate_all(sequences, pfe_methods, network_models, font_directory):
+  """Simulate the matrix of {sequences} x {pfe_methods} x {network_models}.
+
+  For each element compute a set of summary metrics, total time, total
+  request bytes sent, and total response bytes sent.
+  """
+  result = dict()
+  for args in itertools.product(sequences, pfe_methods, network_models):
+    key = "%s (%s)" % (args[1].name(), args[2].name)
```
Comment (rsheeter) [inline analysis/simulation.py:36]: no real reason to stringify, tuples are hashable so just use (name1, name2) as the key directly

File: analysis/simulation.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    data = simulate(args[0], args[1], args[2], font_directory)
```
Comment (rsheeter) [inline analysis/simulation.py:37]: suggest pulling args 0,1,2 into named variables, `for sequenc, pfe_method, net_model in itertools.product...`

File: patch_subset/py/patch_subset_method.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+  def page_viewed(self):
+    self.page_view_count += 1
+    self.records_by_view.append([])
+
+  def extend(self, codepoints):
+    """Extends the tracked font to cover new codepoints.
+
+    Records any resulting requests needed to make the extension.
+    """
```
Comment (rsheeter) [inline patch_subset/py/patch_subset_method.py:90]: say what codepoints is expected to be?

File: patch_subset/py/patch_subset_method.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    codepoint_array_c = (c_uint32 * len(codepoints))()
+    i = 0
+    for codepoint in codepoints:
```
Comment (rsheeter) [inline patch_subset/py/patch_subset_method.py:93]: `for idx, codepoint in enumerate(codepoints): codepoint_array[idx] = codepoint` ?

**Review by rsheeter — APPROVED** [review]

File: analysis/simulation.py (hunk 2 above)
Comment (garretrieger) [reply analysis/simulation.py:36]: This is fixed in a future PR.

File: analysis/simulation.py (hunk 3 above)
Comment (garretrieger) [reply analysis/simulation.py:37]: Fixed in a future change.

File: patch_subset/py/patch_subset_method.py (hunk 4 above)
Comment (garretrieger) [reply patch_subset/py/patch_subset_method.py:90]: Fixed in an upcoming PR.

File: patch_subset/py/patch_subset_method.py (hunk 5 above)
Comment (garretrieger) [reply patch_subset/py/patch_subset_method.py:93]: Fixed in upcoming PR.

File: analysis/analyzer.py (hunk 1 above)
Comment (garretrieger) [reply analysis/analyzer.py:55]: Not sure adding a single type helps, since this function is typically going be called only once per program invocation.

**Pull request merged**